clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java

```diff
@@ -0,0 +1,330 @@
+        if (!coordinatorRequestManager.coordinator().isPresent() || !membershipManager.shouldSendHeartbeat()) {
+            return new NetworkClientDelegate.PollResult(
+                Long.MAX_VALUE, Collections.emptyList());
+        }
+
+        // TODO: We will need to send a heartbeat response after partitions being revoke. This needs to be
+        //  implemented either with or after the partition reconciliation logic.
+        if (!heartbeatRequestState.canSendRequest(currentTimeMs)) {
+            return new NetworkClientDelegate.PollResult(
+                heartbeatRequestState.nextHeartbeatMs(currentTimeMs),
+                Collections.emptyList());
+        }
+        this.heartbeatRequestState.onSendAttempt(currentTimeMs);
+        NetworkClientDelegate.UnsentRequest request = makeHeartbeatRequest();
+        return new NetworkClientDelegate.PollResult(heartbeatRequestState.heartbeatIntervalMs, Collections.singletonList(request));
+    }
+
+    private NetworkClientDelegate.UnsentRequest makeHeartbeatRequest() {
+        // TODO: We only need to send this field once unless the first request failed.
+        ConsumerGroupHeartbeatRequestData data = new ConsumerGroupHeartbeatRequestData()
+            .setGroupId(membershipManager.groupId())
+            .setMemberEpoch(membershipManager.memberEpoch())
+            .setMemberId(membershipManager.memberId())
+            .setRebalanceTimeoutMs(rebalanceTimeoutMs);
+
+        membershipManager.groupInstanceId().ifPresent(data::setInstanceId);
+
+        if (this.subscriptions.hasPatternSubscription()) {
+            // We haven't discsussed how Regex is stored in the consumer. We could do it in the subscriptionState
+            // , in the memberStateManager, or here.
+            // data.setSubscribedTopicRegex(regex)
+        } else {
+            data.setSubscribedTopicNames(new ArrayList<>(this.subscriptions.subscription()));
+        }
+
+        this.membershipManager.assignorSelection().serverAssignor().ifPresent(data::setServerAssignor);
+
+        NetworkClientDelegate.UnsentRequest request = new NetworkClientDelegate.UnsentRequest(
+            new ConsumerGroupHeartbeatRequest.Builder(data),
+            coordinatorRequestManager.coordinator());
+
+        request.future().whenComplete((response, exception) -> {
+            if (exception == null) {
+                onResponse((ConsumerGroupHeartbeatResponse) response.responseBody(), response.receivedTimeMs());
+            } else {
+                onFailure(exception, response.receivedTimeMs());
+            }
+        });
+        return request;
+    }
+
+    private void onFailure(final Throwable exception, final long responseTimeMs) {
+        this.heartbeatRequestState.onFailedAttempt(responseTimeMs);
+        logger.warn("Failed to send heartbeat to coordinator node {} due to error: {}",
+                coordinatorRequestManager.coordinator(), exception.getMessage());
+    }
```

Comment:
My understanding is that we basically retry on all exceptions here. Am I correct? It seems to me that we could also get non-retriable exceptions here (e.g. UnsupportedVersionException, etc.). How do we handle those?

Reply:
I also noticed that we have the following in the current implementation:

```
            if (e instanceof DisconnectException) {
                markCoordinatorUnknown(true, e.getMessage());
            }
```

Reply:
The heartbeat request depends on the state of the state machine, so for most of the non-retriables, it should transition the state to a failure state to prevent additional requests being sent.

The non-retriable also send an error to the handler (`nonRetriableErrorHandler`) to send the error to the user to handle the failure case.

Reply:
Yap I see the DisconnectException handling, it is implemented by all requests going to the coordinator.  Here we handle them independently in the request mangers.  Maybe we could modularize these reaction to make it more unified.

Reply:
@philipnee My point was that the `exception` received here may not be re-triable (e.g. UnsupportedVersionException) and that we may have to take actions on them (e.g. DisconnectException). Are you saying that we handle those in another component? My current understanding is that we don't update the state machine based on those exceptions at the moment because `onErrorResponse` is not called. Do we understand all the exceptions that we could receive here? We need to handle them all appropriately.

Reply:
It seems that we could get the following ones:

```
@Override
        public void onComplete(final ClientResponse response) {
            if (response.authenticationException() != null) {
                onFailure(response.authenticationException());
            } else if (response.wasDisconnected()) {
                onFailure(DisconnectException.INSTANCE);
            } else if (response.versionMismatch() != null) {
                onFailure(response.versionMismatch());
            } else {
                future.complete(response);
            }
        }
```

+ `TimeoutException`

Reply:
I think we might want to refactor the coordinator requests per what you indicated above. A jira is filed: https://issues.apache.org/jira/browse/KAFKA-15531

Otherwise, in the OnFailure() method, I added the logic to fail the state machine and propagate the error when the error is non-retriable